freqtrade/strategy/backtest_lookahead_bias_checker.py

```diff
@@ -0,0 +1,252 @@
+        # starting from the same datetime to avoid miss-reports of bias
+        for idx, result_row in self.full_varHolder.result['results'].iterrows():
+            if self.current_analysis.total_signals == self.targeted_trade_amount:
+                break
+
+            # if force-sold, ignore this signal since here it will unconditionally exit.
+            if result_row.close_date == self.dt_to_timestamp(self.full_varHolder.to_dt):
+                continue
+
+            self.current_analysis.total_signals += 1
+
+            self.entry_varHolder = VarHolder()
+            self.exit_varHolder = VarHolder()
+            self.entry_varHolders.append(self.entry_varHolder)
+            self.exit_varHolders.append(self.exit_varHolder)
+
+            self.entry_varHolder.from_dt = self.full_varHolder.from_dt
+            self.entry_varHolder.compared_dt = result_row['open_date']
+            # to_dt needs +1 candle since it won't buy on the last candle
+            self.entry_varHolder.to_dt = (result_row['open_date'] +
+                                          timedelta(minutes=timeframe_to_minutes(
+                                              self.local_config['timeframe'])))
+
+            self.prepare_data(self.entry_varHolder, [result_row['pair']])
+
+            # to_dt needs +1 candle since it will always exit/force-exit trades on the last candle
+            self.exit_varHolder.from_dt = self.full_varHolder.from_dt
+            self.exit_varHolder.to_dt = (result_row['close_date'] +
+                                         timedelta(minutes=timeframe_to_minutes(
+                                             self.local_config['timeframe'])))
+            self.exit_varHolder.compared_dt = result_row['close_date']
+
+            self.prepare_data(self.exit_varHolder, [result_row['pair']])
+
+            # register if buy signal is broken
+            if not self.report_signal(
+                    self.entry_varHolder.result, "open_date", self.entry_varHolder.compared_dt):
+                self.current_analysis.false_entry_signals += 1
+
+            # register if buy or sell signal is broken
+            if not self.report_signal(
+                    self.exit_varHolder.result, "close_date", self.exit_varHolder.compared_dt):
+                self.current_analysis.false_exit_signals += 1
+
+            if len(self.entry_varHolders) >= 10:
+                pass
+            # check if the indicators themselves contain biased data
+            self.analyze_indicators(self.full_varHolder, self.entry_varHolder, result_row['pair'])
+            self.analyze_indicators(self.full_varHolder, self.exit_varHolder, result_row['pair'])
+
+        if (self.current_analysis.false_entry_signals > 0 or
+                self.current_analysis.false_exit_signals > 0 or
+                len(self.current_analysis.false_indicators) > 0):
+            print(" => " + self.local_config['strategy_list'][0] + ": bias detected!")
```

Comment:
i assume the usage of `print` here is intentional?

in general, i'd use logging instead unless there's good reasons against it.
Assuming this can run for a longer period - it'll allow implicit piping to file (`--log-file`) - while print output will be to stdout - which will be gone if you close the terminal (and didn't redirect output).

At any rate, i'd prefer f-strings instead for string concatenation - they feel more modern (and have some implicit type-conversion that can help, too).

Reply:
switched to f-strings and logging.info
yes, f-strings are much better.